trieste/space.py

```diff
@@ -135,44 +139,47 @@ def __mul__(self, other: DiscreteSearchSpace) -> DiscreteSearchSpace:
                    [2, 3, 4, 5, 6],
                    [2, 3, 7, 8, 9]], dtype=int32)
 
-        :param other: :class:`DiscreteSearchSpace`.
-        :return: the new combined :class:`DiscreteSearchSpace`
-        :raise TypeError: If the lhs and rhs :class:`DiscreteSearchSpace` points have different
-            types.
+        :param other: A :class:`DiscreteSearchSpace` with :attr:`points` of the same dtype as this
+            search space.
+        :return: The Cartesian product of the two :class:`DiscreteSearchSpace`\ s.
+        :raise TypeError: If one :class:`DiscreteSearchSpace` has :attr:`points` of a different
+            dtype to the other.
         """
-        if self._points.dtype is not other._points.dtype:
+        if self.points.dtype is not other.points.dtype:
             return NotImplemented
 
-        N = self._points.shape[0]
-        M = other._points.shape[0]
-        tile_self = tf.tile(tf.expand_dims(self.points, 1), [1, M, 1])
-        tile_dss = tf.tile(tf.expand_dims(other.points, 0), [N, 1, 1])
-        cartesian_product = tf.concat([tile_self, tile_dss], axis=2)
-
-        return DiscreteSearchSpace(tf.reshape(cartesian_product, [N * M, -1]))
+        tile_self = tf.tile(self.points[:, None], [1, len(other.points), 1])
```

Comment:
this works only with rank 2 tensors; are we happy with this / asserting the ranks in the first place?

Reply:
the class `__init__` asserts rank 2

Reply:
(which I'll mention isn't a long term plan - just until we've made it all work with higher rank)